packages/vite/src/node/server/shortcuts.ts

```diff
@@ -2,30 +2,33 @@ import type { ViteDevServer } from '..'
 import { openBrowser, resolveBrowserUrl } from './openBrowser'
 import { restartServer } from './hmr'
 
-export function bindShortcuts(server: ViteDevServer, isRestart = false): void {
+export function bindShortcuts(server: ViteDevServer): void {
+  if (!server.httpServer) return
```

Comment:
Since we use `httpServer.on('close', ...)` to unbind the shortcuts, we can't bind the shortcuts without it. This means middleware mode is off limits.